src/VisualStudio/Core/Def/Implementation/Watson/WatsonExtensions.cs

```diff
@@ -0,0 +1,78 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using Microsoft.VisualStudio.Telemetry;
+using Roslyn.Utilities;
+using StreamJsonRpc;
+
+namespace Microsoft.CodeAnalysis.ErrorReporting
+{
+    internal static class WatsonExtensions
+    {
+        /// <summary>
+        /// This sets extra watson bucket parameters to make bucketting better
+        /// in non fatal watson report
+        /// </summary>
+        public static void SetExtraParameters(this IFaultUtility fault, Exception exception)
+        {
+            switch (exception)
+            {
+                case RemoteInvocationException remote:
+                    fault.SetBucketParameter(8, remote.GetParameterString());
+                    return;
+                case AggregateException aggregate:
+                    if (aggregate.InnerException == null)
+                    {
+                        return;
+                    }
+                    else if (aggregate.InnerExceptions.Count == 1)
+                    {
+                        fault.SetBucketParameter(8, aggregate.GetParameterString());
+                        return;
+                    }
+                    else
+                    {
+                        var flatten = aggregate.Flatten();
+
+                        fault.SetBucketParameter(7, flatten.CalculateHash());
+                        fault.SetBucketParameter(8, flatten.InnerException.GetParameterString());
+                        return;
+                    }
+                default:
+                    if (exception.InnerException == null)
+                    {
+                        return;
+                    }
+
+                    fault.SetBucketParameter(8, exception.InnerException.GetParameterString());
+                    return;
+            }
+        }
+
+        public static string CalculateHash(this AggregateException exception)
+        {
+            var hash = 1;
+            foreach (var inner in exception.InnerExceptions)
+            {
+                var parameterString = inner.GetParameterString();
+                hash = Hash.Combine(parameterString, hash);
+            }
+
+            return hash.ToString();
+        }
+
+        public static string GetParameterString(this Exception exception)
+        {
+            switch (exception)
+            {
+                case RemoteInvocationException remote:
+                    return $"{remote.RemoteErrorCode} {remote.RemoteStackTrace ?? exception.Message}";
```

Comment:
this is not enough for remoteInvocationException since servicehub itself disconnection issue we have had will have canned info in there. we need even more data to distinguish for those such as service hub log and etc. that will be another improvement we can do later.

but this should still help some cases and better than before.